fenwick/src/main/java/org/opencadc/fenwick/InventoryHarvester.java

```diff
@@ -185,43 +194,108 @@ public InventoryHarvester(Map<String, Object> daoConfig, URI resourceID, Artifac
-                long dt = 60 * 1000L;
-                Thread.sleep(dt);
-            } catch (PrivilegedActionException privilegedActionException) {
-                final Exception exception = privilegedActionException.getException();
-                throw new IllegalStateException(exception.getMessage(), exception);
-            } catch (IllegalArgumentException ex) {
-                // Be careful here.  This IllegalArgumentException is being caught to work around a mysterious
-                // case where TCP connections are simply dropped and the incoming stream of data is invalid.
-                // This catch will allow Fenwick to restart its processing.
-                // jenkinsd 2020.09.25
-                final String message = ex.getMessage().trim();
-                if (message.startsWith("wrong number of columns")) {
-                    log.error("\n\n*******\n");
-                    log.error("Caught IllegalArgumentException - " + message + " (" + ++errorCount + ")");
-                    log.error("Ignoring error as presumed to be a dropped connection before fully reading the stream.");
-                    log.error("\n*******\n");
-                } else if (message.startsWith("invalid checksum URI:")) {
-                    log.error("\n\n*******\n");
-                    log.error("Caught IllegalArgumentException - " + message + " (" + ++errorCount + ")");
-                    log.error("Ignoring error as presumed to be a dropped connection before fully reading the stream.");
-                    log.error("CAUTION! - This could actually be a bad MD5 checksum! Logging this to provide an audit.");
-                    log.error("\n*******\n");
+                if (retry) {
+                    Thread.sleep(sleepSeconds * 1000L);
+                    retryCount++;
+                    if (sleepSeconds * 2 < this.maxRetryInterval) {
+                        sleepSeconds *= 2;
+                        if (sleepSeconds > this.maxRetryInterval) {
+                            sleepSeconds = this.maxRetryInterval;
+                        }
+                    }
                 } else {
-                    throw ex;
+                    // successful run, reset retry values
+                    retryCount = 1;
+                    sleepSeconds = INITIAL_SLEEP_TIMEOUT;
+                    // TODO: dynamic depending on how rapidly the remote content is changing
+                    // ... this value and the reprocess-last-N-seconds should be related
+                    Thread.sleep(DEFAULT_SLEEP_TIMEOUT * 1000L);
                 }
             } catch (InterruptedException ex) {
-                throw new RuntimeException("interrupted", ex);
+                // Thread interrupted, fail.
+                logExit("Thread interrupted", ex.getMessage());
+                throw new RuntimeException(ex.getMessage(), ex);
+            } catch (NotAuthenticatedException ex) {
```

Comment:
The InterruptedException should not be caught and wrapped. Just let it fly.

The NotAuthenticatedException from HttpTransfer if it thinks the proxy cert in the subject was rejected by the server (eg expired), so this is actually a failure that could start working if ops updates the cert in the config dir. So technically recoverable without restarting fenwick.

The AccessControlException from HttpTransfer will be due to permissions at the remote luskan and could start working is ops fixes some other config (usually baldur or luskan) or replaces the proxy cert with the right one. So technically recoverable without restarting fenwick.

Reply:
The InterruptedException is a checked exception in the run method that has to be caught.

I asked myself if a certificate expired, what would ops notice first: an error message in a log, or the fact that fenwick had stopped running. 

If it's expected that ops will detect and fix cert issues, then should retry on cert exceptions, updated.